### PR TITLE
Features/annotation

### PR DESCRIPTION
**Pull request by gupele**

### CODE REVIEW

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java [hunk 1]
```diff
@@ -40,9 +37,12 @@
 import com.microsoft.applicationinsights.extensibility.initializer.SdkVersionContextInitializer;
 import com.microsoft.applicationinsights.internal.annotation.AnnotationPackageScanner;
 import com.microsoft.applicationinsights.internal.annotation.PerformanceModule;
+import com.microsoft.applicationinsights.internal.jmx.JmxAttributeData;
 import com.microsoft.applicationinsights.internal.logger.InternalLogger;
 
 import com.google.common.base.Strings;
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: reorder

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java [hunk 2]
```diff
@@ -183,17 +183,56 @@ private boolean setInstrumentationKey(String instrumentationKey, TelemetryConfig
                 new AnnotationPackageScanner().scanForClassAnnotations(new Class[]{PerformanceModule.class}, performanceCountersSection);
         for (String performanceModuleName : performanceModuleNames) {
             TelemetryModule module = createInstance(performanceModuleName, TelemetryModule.class);
-//            PerformanceModule pmAnnotation = module.getClass().getAnnotation(PerformanceModule.class);
             if (module != null) {
+                PerformanceModule pmAnnotation = module.getClass().getAnnotation(PerformanceModule.class);
+                if (!performanceConfigurationData.isUseBuiltIn() && "BuiltIn".equals(pmAnnotation.value())) {
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: use consts for BuiltIn (used also in the annotation impl)

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java [hunk 3]
```diff
@@ -183,17 +183,56 @@ private boolean setInstrumentationKey(String instrumentationKey, TelemetryConfig
...
 
+    private void loadCustomJmxPCs(ArrayList<JmxXmlElement> jmxXmlElements) {
+        try {
+            if (jmxXmlElements == null) {
+                return;
+            }
+
+            HashMap<String, Collection<JmxAttributeData>> data = new HashMap<String, Collection<JmxAttributeData>>();
+
+            for (JmxXmlElement jmxElement : jmxXmlElements) {
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: please add a comment to explain what you're doing here.

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java [hunk 4]
```diff
@@ -371,7 +410,7 @@ void setPerformanceCountersSection(String performanceCountersSection) {
         this.performanceCountersSection = performanceCountersSection;
     }
 
-    void setBuilder(AppInsightsConfigurationReader builder) {
+    void setBuilder(AppInsightsConfigurationBuilder builder) {
         this.builder = builder;
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: why this method if you only set?

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryInitializersXmlElement.java [hunk 5]
```diff
@@ -41,4 +42,13 @@
     public void setAdds(ArrayList<AddTypeXmlElement> adds) {
         this.adds = adds;
     }
+
+    public ArrayList<AddTypeXmlElement> getRemoves() {
+        return removes;
+    }
+
+    @XmlElement(name="Remove")
+    public void setRemoves(ArrayList<AddTypeXmlElement> removes) {
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryInitializersXmlElement.java]: RemoveTypeXmlElement ?

File: core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxAttributeData.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+ * Represents JMX data of an Attribute
+ * The display name
+ * The name of the attribute
+ * The type of the attribute
+ *
+ * Created by gupele on 3/15/2015.
+ */
+public final class JmxAttributeData {
+    public final String displayName;
+    public final String name;
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxAttributeData.java]: name of what? object name?

Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxAttributeData.java]: attributeName

File: core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxDataFetcher.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+public class JmxDataFetcher {
+    /**
+     * Gets an object name and its attributes to fetch and will return the data.
+     * @param objectName The object name to search.
+     * @param attributes The attributes that 'belong' to the object name.
+     * @return A map that represent each attribute: the key is the displayed name for that attribute
+     * and the value is a list of values found
+     * @throws Exception In case the object name is not found.
+     */
+    public static Map<String, Collection<Object>> fetch(String objectName, Collection<JmxAttributeData> attributes) throws Exception {
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxDataFetcher.java]: I think it should throw any exception.

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java (hunk 4 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: for testing that's why it is package level access

File: core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxDataFetcher.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        Map<String, Collection<Object>> result = new HashMap<String, Collection<Object>>();
+
+        MBeanServer server = ManagementFactory.getPlatformMBeanServer();
+        Set<ObjectName> objects = server.queryNames(new ObjectName(objectName), null);
+        if (objects.isEmpty()) {
+            String errorMsg = String.format("Cannot find object name '%s'", objectName);
+            throw new IllegalArgumentException(errorMsg);
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxDataFetcher.java]: simply return null

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryInitializersXmlElement.java (hunk 5 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryInitializersXmlElement.java]: Yes, ready for adding <Remove type=""/> in our configuration

File: core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxDataFetcher.java (hunk 7 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxDataFetcher.java]: The reason it throws is that if there is a major problem like the object name is not ok, then there is no point of trying again

File: core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxDataFetcher.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+        for (JmxAttributeData attribute : attributes) {
+            try {
+                Collection<Object> resultForAttribute = fetch(server, objects, attribute.name);
+                result.put(attribute.displayName, resultForAttribute);
+            } catch (Exception e) {
+                InternalLogger.INSTANCE.error("Failed to fetch JMX attribute %s, %s: '%s'", objectName, attribute.name, e.getMessage());
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxDataFetcher.java]: "Failed to fetch JMX object 'x' with attribute 'y'."

File: core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxDataFetcher.java (hunk 8 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxDataFetcher.java]: This is major error where the object name is not legal, so there is a point in the exception so the caller knows that

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java (hunk 2 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: done

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractJmxPerformanceCounter.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+                        ok = false;
+                        break;
+                    }
+                }
+
+                if (ok) {
+                    send(telemetryClient, displayAndValues.getKey(), value);
+                }
+            }
+        } catch (Exception e) {
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractJmxPerformanceCounter.java]: No specific exception to identify that the object name no found?

Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractJmxPerformanceCounter.java]: In addition, I think it worth doing this inside an "Initialize" method, makes more sense.

Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractJmxPerformanceCounter.java]: In addition, please explain why you're ignoring on failure on the first time (you're assuming that it is not exists

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractJmxPerformanceCounter.java [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+                for (Object obj : displayAndValues.getValue()) {
+                    try {
+                        value += Double.parseDouble(String.valueOf(obj));
+                    } catch (Exception e) {
+                        ok = false;
+                        break;
+                    }
+                }
+
+                if (ok) {
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractJmxPerformanceCounter.java]: will fail on firstTime

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java (hunk 1 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: done

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java (hunk 3 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: done

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractPerformanceCounterBase.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+package com.microsoft.applicationinsights.internal.perfcounter;
+
+import com.microsoft.applicationinsights.internal.system.SystemInformation;
+
+/**
+ * A base class for classes that report performance counters of Process level.
+ *
+ * Created by gupele on 3/12/2015.
+ */
+abstract class AbstractPerformanceCounterBase implements PerformanceCounter {
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractPerformanceCounterBase.java]: AbstractProcessPerformanceCounter

File: core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxDataFetcher.java (hunk 9 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/jmx/JmxDataFetcher.java]: done

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractPerformanceCounterModuleBase.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
+/*
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractPerformanceCounterModuleBase.java]: remove Base

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractJmxPerformanceCounter.java (hunk 11 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractJmxPerformanceCounter.java]: done

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractJmxPerformanceCounter.java (hunk 10 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractJmxPerformanceCounter.java]: done

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractPerformanceCounterModuleBase.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+     * @param configuration The configuration to used to initialize the module.
+     */
+    @Override
+    public void initialize(TelemetryConfiguration configuration) {
+        Collection<PerformanceCounter> performanceCounters = factory.getPerformanceCounters();
+        for (PerformanceCounter performanceCounter : performanceCounters) {
+            try {
+                PerformanceCounterContainer.INSTANCE.register(performanceCounter);
+            } catch (Throwable e) {
+                InternalLogger.INSTANCE.error("Failed to register performance counter: '%s'", e.getMessage());
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractPerformanceCounterModuleBase.java]: add the name of the perf counter

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractUnixPerformanceCounterBase.java [hunk 15]
```diff
@@ -0,0 +1,60 @@
+/*
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractUnixPerformanceCounterBase.java]: remove Base

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractPerformanceCounterModuleBase.java (hunk 13 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractPerformanceCounterModuleBase.java]: done

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractPerformanceCounterModuleBase.java (hunk 14 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractPerformanceCounterModuleBase.java]: done

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractUnixPerformanceCounterBase.java (hunk 15 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractUnixPerformanceCounterBase.java]: done

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractPerformanceCounterModuleBase.java (hunk 14 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/AbstractPerformanceCounterModuleBase.java]: done

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/Constants.java [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    public final static String PROCESS_MEM_PC_COUNTER_NAME = "Private Bytes";
+
+    public final static String TOTAL_MEMORY_PC_CATEGORY_NAME = "Memory";
+    public final static String TOTAL_MEMORY_PC_COUNTER_NAME = "Available Bytes";
+
+
+    public final static String PROCESS_IO_PC_ID = "JSDK_ProcessIOPerformanceCounter";
+    public final static String PROCESS_IO_PC_COUNTER_NAME = "IO Data Bytes/sec";
+
+    public static final double  MEGA_BYTE = 1024L * 1024L;
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/Constants.java]: redundant space

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/Constants.java [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+    public final static String PERFORMANCE_COUNTER_PREFIX = "JSDK_";
+
+    public final static String TOTAL_CPU_PC_ID = "JSDK_TotalCpuPerformanceCounter";
+
+    public final static String PROCESS_CPU_PC_ID = "JSDK_ProcessCpuPerformanceCounter";
+
+    public final static String TOTAL_CPU_PC_CATEGORY_NAME = "Processor(_Total)";
+    public final static String CPU_PC_COUNTER_NAME = "% Processor Time";
+
+
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/Constants.java]: remove line

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/Constants.java (hunk 16 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/Constants.java]: done

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/Constants.java [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+package com.microsoft.applicationinsights.internal.perfcounter;
+
+/**
+ * Created by gupele on 3/8/2015.
+ */
+public final class Constants {
+    public final static String PERFORMANCE_COUNTER_PREFIX = "JSDK_";
+
+    public final static String TOTAL_CPU_PC_ID = "JSDK_TotalCpuPerformanceCounter";
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/Constants.java]: use PERFORMANCE_COUNTER_PREFIX. for below as well

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/PerformanceCounterContainer.java [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+ *
+ * The container will be stopped automatically when the application exists.
+ *
+ * Created by gupele on 3/3/2015.
+ */
+public enum PerformanceCounterContainer implements Stoppable {
+    INSTANCE;
+
+    // By default the container will wait 2 minutes before the collection of performance data.
+    private final static long START_COLLECTING_INTERVAL_IN_MILLIS = 10000;// 120000;
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/PerformanceCounterContainer.java]: 120000

Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/PerformanceCounterContainer.java]: START_COLLECTION_DELAY_IN_MILLIS

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/PerformanceCounterContainer.java [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+
+    // By default the container will collect performance data every 1 minute.
+    private final static long COLLECTING_INTERVAL_IN_MILLIS = 60000;
+
+    private final ConcurrentMap<String, PerformanceCounter> performanceCounters = new ConcurrentHashMap<String, PerformanceCounter>();
+
+    private volatile boolean initialized = false;
+
+    private long startCollectingIntervalInMillis = START_COLLECTING_INTERVAL_IN_MILLIS;
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/PerformanceCounterContainer.java]: startCollectionDelay...

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/PerformanceCounterContainer.java [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+    private long collectingIntervalInMillis = COLLECTING_INTERVAL_IN_MILLIS;
+
+    private TelemetryClient telemetryClient;
+
+    private ScheduledThreadPoolExecutor threads;
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/PerformanceCounterContainer.java]: threadPool

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/ProcessBuiltInPerformanceCountersFactory.java [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+ * The class will create the 'built-in'/default performance counters.
+ *
+ * Created by gupele on 3/3/2015.
+ */
+final class ProcessBuiltInPerformanceCountersFactory implements PerformanceCountersFactory {
+    /**
+     * Creates the {@link com.microsoft.applicationinsights.internal.perfcounter.PerformanceCounter} that are
+     * the 'built-in' performance counters of the process.
+     *
+     * Note: The method should not throw
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/ProcessBuiltInPerformanceCountersFactory.java]: Remove

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/ProcessBuiltInPerformanceCountersFactory.java [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+        performanceCounters.add(new ProcessMemoryPerformanceCounter());
+        performanceCounters.add(new ProcessCpuPerformanceCounter());
+
+        return performanceCounters;
+    }
+
+    /**
+     * Returns the performance counters that are useful for Windows OS.
+     * @return Collection of Windows performance counters.
+     */
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/ProcessBuiltInPerformanceCountersFactory.java]: fix the comments - not Windows --> Unix

Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/ProcessBuiltInPerformanceCountersFactory.java]: Simply switch with the Windows method (which you put there comment for Unix) :)

Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/ProcessBuiltInPerformanceCountersFactory.java]: done

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/ProcessBuiltInPerformanceCountersFactory.java (hunk 22 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/ProcessBuiltInPerformanceCountersFactory.java]: done

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/UnixProcessIOPerformanceCounter.java [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+
+    @Override
+    public void report(TelemetryClient telemetryClient) {
+        long currentCollectionInNanos = System.nanoTime();
+
+        if (lastCollectionInNanos != -1) {
+            // Not the first time
+
+            double processIO = getCurrentIOForCurrentProcess();
+            double timeElapsedInSeconds = ((double)(currentCollectionInNanos - lastCollectionInNanos)) / 1000000000.0;
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/UnixProcessIOPerformanceCounter.java]: const

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/UnixProcessIOPerformanceCounter.java [hunk 25]
```diff
@@ -0,0 +1,134 @@
...
+                }
+                if (!writeBytesDone) {
+                    writeBytesDone = parseValue(parsingData, line, WRITE_BYTES_PART);
+                }
+            }
+
+            result = parsingData.returnValue;
+        } catch (Exception e) {
+            result = Constants.DEFAULT_DOUBLE_VALUE;
+            logError("Error while parsing file: '%s'", e.getMessage());
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/UnixProcessIOPerformanceCounter.java]: Add indication that it was during processing of X performance counter

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/UnixTotalCpuPerformanceCounter.java [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    @Override
+    public void report(TelemetryClient telemetryClient) {
+        String line = getLineOfData();
+
+        if (!Strings.isNullOrEmpty(line)) {
+            String[] rawStringValues = line.split(" ");
+
+            ArrayList<String> stringValues = new ArrayList<String>(rawStringValues.length - 1);
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/UnixTotalCpuPerformanceCounter.java]: why -1

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/UnixTotalCpuPerformanceCounter.java [hunk 27]
```diff
@@ -0,0 +1,145 @@
...
+                    continue;
+                }
+
+                stringValues.add(stringValue);
+            }
+
+            String[] array = stringValues.toArray(new String[stringValues.size()]);
+
+            if (prevCpuCounters == null) {
+                firstCounters(array);
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/UnixTotalCpuPerformanceCounter.java]: please rename, it doesn't reflect what the method does

Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/UnixTotalCpuPerformanceCounter.java]: done

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/UnixProcessIOPerformanceCounter.java (hunk 24 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/UnixProcessIOPerformanceCounter.java]: done

File: core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/UnixProcessIOPerformanceCounter.java (hunk 25 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/UnixProcessIOPerformanceCounter.java]: done, added at base

File: core/src/main/java/com/microsoft/applicationinsights/internal/shutdown/SDKShutdownActivity.java [hunk 28]
```diff
@@ -92,6 +99,19 @@ private void stopChannels() {
...
+        }
+        /**
+         * Make sure no exception is thrown!
+         */
+        private void stopStoppables() {
+            for (Stoppable stoppable : stoppables) {
+                try {
+                    stoppable.stop(1L, TimeUnit.SECONDS);
+                } catch (Throwable t) {
+                    InternalLogger.INSTANCE.error("Failed to stop stoppable: '%s'", t.getMessage());
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/shutdown/SDKShutdownActivity.java]: add the name of the class failed to be stopped

File: core/src/main/java/com/microsoft/applicationinsights/internal/system/SystemInformation.java [hunk 29]
```diff
@@ -0,0 +1,89 @@
...
+        if (!Strings.isNullOrEmpty(rawName)) {
+            int i = rawName.indexOf("@");
+            if (i != -1) {
+                String processIdAsString = rawName.substring(0, i);
+                try {
+                    Integer.parseInt(processIdAsString);
+                    processId = processIdAsString;
+                    return;
+                } catch (Exception e) {
+                }
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/system/SystemInformation.java]: error

File: core/src/main/java/com/microsoft/applicationinsights/internal/system/SystemInformation.java [hunk 30]
```diff
@@ -0,0 +1,89 @@
...
+//            jvmField.setAccessible(true);
+//            VMManagement vmManagement = (VMManagement) jvmField.get(runtimeMXBean);
+//            Method getProcessIdMethod = vmManagement.getClass().getDeclaredMethod("getProcessId");
+//            getProcessIdMethod.setAccessible(true);
+//            Integer pId = (Integer) getProcessIdMethod.invoke(vmManagement);
+//            processId = String.valueOf(pId);
+//
+//            return;
+//        } catch (Exception e) {
+//        }
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/system/SystemInformation.java]: remove

**Pull request merged**